client/DESIGN.md

```diff
@@ -0,0 +1,380 @@
+
+HTTP read requests should follow this format:
+
+    GET /[table_id_string]/[key_string] HTTP/1.1
+    Cache-Control: max-age=[max_age_in_seconds], no-cache
+    Read-TxClock: [read_timestamp_in_microseconds_long_64_bits]
+
+Here is an example of an HTTP read request:
+
+    GET /movie/star-wars HTTP/1.1
+    Cache-Control: max-age=300, no-cache
+    Read-TxClock: 1421236153024853
+
+The user is requesting the value of key "star-wars" in the table "movie".  Since this key's value may have changed over time, the user gives a time range for the value they want.  
+
+Specifically, the user wants a value that was written to the DB by time 1421236153024853 microseconds (where time 0 is some arbitrary starting point in the past).  However, the user does not want a value that was written to the DB more than 300 seconds before time 1421236153024853.  Thus, the time range is 
+
+    [1421236153024853 - 300, 1421236153024853].  
+
+Finally, the "no-cache" indicates that the cache *must* send a read request to the DB server to retrieve this value, even if the cache already has a value satisfying the user's requirements.
+
+#### Write Requests  
+
+The cache message to the DB includes the table id and a list of operations to perform on the table.  The operations should be sent to the DB server as a Treode batch write operation.
+
+The HTTP request for write should be similar to the following:
+
+    PUT /table_id_string/key_string HTTP/1.1
+    Condition-TxClock: [condition_timestamp_in_milliseconds]
+    {
+        [JSON_value]
+    }
+
+Here is an example of the HTTP write request:
+
+    PUT /movie/star-wars HTTP/1.1
+    Condition-TxClock: 1368151663681367
+    {   "title": "Star Wars",
+        "cast": [
+            { "actor": "Mark Hamill", "role": "Luke Skywalker" },
+            { "actor": "Carrie Fisher", "role": "Princess Leia Organa" },
+            { "actor": "Harrison Ford", "role": "Han Solo" } ] 
+    }
+
+We write the given JSON content to key "star-wars" in table "movie", provided that the current value associated with the "star-wars" key has remained unchanged since the time 1368151663681367 milliseconds.
+
+In addition, the DB server supports batch writes, and the cache will in prctice use batch writes instead of the traditional write format described above.
+
+There are several batch write formats the DB server accepts:
+
+##### JSON Array
+Simply wrap the multiple write requests in a JSON array, as shown below.
+
+    [ { "op": "create", "table": "movie", "key": "star-wars", value={ /* movie object */ } },
+      { "op": "delete", "table": "movie", "key": "stars-war" } } ]
+
+The rest of the HTTP request remains as described above.
+
+##### Request Pipelining
+
+Send each item in the batch write to the server as a separate HTTP request.  We indicate in the requests that they are part of a batch write with the item header as shown below in the example two-part batch write:
+
+    PUT /movie/star-wars HTTP/1.1
+    X-Item: 1/2
```

Comment:
Let's make the header like

    Transaction: id=X, item=1/2

The `id` is optional and the server will fill in something. It is a hex encoding of bytes.

The `item` is also optional, and the server will fill in 1/1.

Yes, we can make up our own headers. That's why HTTP is cool.